python/ray/data/_internal/execution/operators/hash_shuffle.py

```diff
@@ -445,12 +446,143 @@ def __init__(
+                f"Required: {required_cpus} CPUs for {self._aggregator_pool.num_aggregators} aggregators, "
+                f"but cluster only has {total_cpus} total CPUs. "
+                f"Consider reducing the number of partitions or increasing cluster size."
+            )
+
+        if required_cpus > available_cpus:
+            logger.warning(
+                f"Limited available CPU resources for hash shuffle operation. "
+                f"Required: {required_cpus} CPUs, available: {available_cpus} CPUs. "
+                f"Aggregators may take longer to start due to resource contention."
+            )
+
+        # Check memory resources if specified
+        if required_memory > 0:
+            total_memory = cluster_resources.get("memory", 0)
+            available_memory = available_resources.get("memory", 0)
+
+            if required_memory > total_memory:
+                raise ValueError(
+                    f"Insufficient memory resources in cluster for hash shuffle operation. "
+                    f"Required: {required_memory / GiB:.2f} GiB for {self._aggregator_pool.num_aggregators} aggregators, "
+                    f"but cluster only has {total_memory / GiB:.2f} GiB total memory. "
+                    f"Consider reducing the number of partitions or increasing cluster size."
+                )
+
+            if required_memory > available_memory:
+                logger.warning(
+                    f"Limited available memory resources for hash shuffle operation. "
+                    f"Required: {required_memory / GiB:.2f} GiB, available: {available_memory / GiB:.2f} GiB. "
+                    f"Aggregators may take longer to start due to resource contention."
+                )
+
+        logger.debug(
+            f"Resource check passed for hash shuffle operation: "
+            f"required CPUs={required_cpus}, available CPUs={available_cpus}, "
+            f"required memory={required_memory / GiB:.2f} GiB, available memory={available_memory / GiB:.2f} GiB"
+        )
+
+    def _check_aggregator_health(self, timeout_seconds: float = 30.0) -> None:
+        """Check if all aggregators are up and running after a timeout period.
+
+        Args:
+            timeout_seconds: Time to wait before checking aggregator health.
+        """
+        if (
+            self._aggregator_startup_time is None
+            or self._aggregator_monitoring_warned
+            or time.time() - self._aggregator_startup_time < timeout_seconds
+        ):
+            return
+
+        try:
+            # Check if all aggregators are responsive
+            unresponsive_aggregators = []
+
+            for i, aggregator in enumerate(self._aggregator_pool._aggregators):
+                try:
+                    # Try to get a simple health check from the actor
+                    # We use a very short timeout to avoid blocking
+                    ray.get(aggregator.__ray_ready__.remote(), timeout=1.0)
```

Comment:
This could block for arbitrarily long for large number of actors.

Instead let's do this 

 - Take a look at ActorPoolMapOperator how it's handling actors readiness
 - We invoke remote methods on all the actors handlers
 - We poll on them using ray.wait (in unblocking fashion) until all become ready
 - When all become ready that means all actors have successfully started